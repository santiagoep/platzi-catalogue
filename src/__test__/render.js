/* istanbul ignore file */
/* eslint-disable import/no-extraneous-dependencies */
import { Provider } from 'react-redux';
import { mount, shallow } from 'enzyme';
import Renderer from 'react-test-renderer';
import { render } from '@testing-library/react';
import { ThemeProvider } from 'styled-components';

import { makeMockedStore } from '@store/index';
import mockedStore from '@__mocks__/store/index';
import lightTheme from '@assets/styles/themes/light';

const WrapperComponent = ({ children, config, initialState }) => {
  if (config && config.theme && config.redux) {
    const store = makeMockedStore(initialState || mockedStore);
    return (
      <Provider store={store}>
        <ThemeProvider theme={lightTheme}>{children}</ThemeProvider>
      </Provider>
    );
  }
  if (config && config.redux) {
    const store = makeMockedStore(initialState || mockedStore);
    return <Provider store={store}>{children}</Provider>;
  }
  if (config && config.theme) {
    return <ThemeProvider theme={lightTheme}>{children}</ThemeProvider>;
  }
  return children;
};

const renderWithProviders = (
  { redux, theme } = {},
  initialState,
  getMockedStore
) => {
  const renderProviders = (Component) => (
    <WrapperComponent
      config={{ redux, theme }}
      initialState={initialState}
      getMockedStore={getMockedStore}
    >
      {Component}
    </WrapperComponent>
  );

  return {
    mount: (Component) => mount(renderProviders(Component)),
    render: (Component) => render(renderProviders(Component)),
    shallow: (Component) => shallow(renderProviders(Component)),
    renderer: (Component) => Renderer.create(renderProviders(Component))
  };
};

export default renderWithProviders;
