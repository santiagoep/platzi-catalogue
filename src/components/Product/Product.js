import PropTypes from 'prop-types';

import Link from '@components/Link/Link';
import formatCurrency from '@utils/numbers';
import FancyLine from '@components/FancyLine/FancyLine';
import Typography from '@components/Typography/Typography';
import FakeButton from '@components/FakeButton/FakeButton';
import ProductImage from '@components/ProductImage/ProductImage';
import { StyledProductImageContainer, StyledProduct } from './Product.styled';

const Product = ({ slug, name, price, image, hoverImage, onClick }) => (
  <Link href={`/products/${slug}`} onClick={onClick}>
    <StyledProduct>
      <StyledProductImageContainer>
        <ProductImage {...hoverImage} className='bottom' />
        <ProductImage {...image} className='top' />
      </StyledProductImageContainer>
      <Typography
        type='subtitle1'
        color='secondary'
        fontWeight='bold'
        textAlign='center'
        as='h2'
      >
        {name}
      </Typography>
      <Typography
        type='subtitle1'
        color='primary'
        fontWeight='bold'
        textAlign='center'
        as='h2'
      >
        {formatCurrency(price)}
      </Typography>
      <FancyLine />
      <FakeButton centered>
        <Typography color='tertiary' fontWeight='bold'>
          Ver m√°s
        </Typography>
      </FakeButton>
    </StyledProduct>
  </Link>
);

Product.propTypes = {
  slug: PropTypes.string,
  name: PropTypes.string,
  price: PropTypes.number,
  image: PropTypes.shape({
    src: PropTypes.string,
    alt: PropTypes.string
  }),
  hoverImage: PropTypes.shape({
    src: PropTypes.string,
    alt: PropTypes.string
  }),
  onClick: PropTypes.func
};

Product.defaultProps = {
  slug: '/',
  name: '',
  price: undefined,
  image: {},
  hoverImage: undefined,
  onClick: () => {}
};

export default Product;
