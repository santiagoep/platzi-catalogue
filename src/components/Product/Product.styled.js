import styled from 'styled-components';

export const StyledProductImageContainer = styled.div`
  position: relative;
  height: 50vh;
  width: auto;
  img {
    transition: opacity 0.8s ease-in-out;
  }
  img.top:hover {
    opacity: 0;
  }
`;

export const StyledProduct = styled.div`
  text-align: center;
  margin: ${({ theme }) => theme.utils.spacing(5)};
  padding: ${({ theme }) => theme.utils.spacing(5)};
  box-shadow: ${({ theme }) => theme.boxShadows.main};
  border-radius: ${({ theme }) => theme.borderRadius.medium};
  background-color: ${({ theme }) => theme.colors.bg.secondary};
  :hover {
    box-shadow: ${({ theme }) => theme.boxShadows.secondary};
  }
`;
