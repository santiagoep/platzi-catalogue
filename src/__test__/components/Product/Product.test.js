import render from '../../render';
import productMock from '@__mocks__/components/product';
import Product from '@components/Product/Product';

describe('<Product />', () => {
  it('Should render snapshot', () => {
    const { mount } = render({ theme: true });
    const component = mount(<Product {...productMock} />);
    expect(component).toMatchSnapshot();
  });
  it('Should click on product', () => {
    const { mount } = render({ theme: true });
    const spy = jest.fn();
    const component = mount(<Product onClick={spy} {...productMock} />);
    component.simulate('click');
    expect(spy).toHaveBeenCalled();
  });
});
