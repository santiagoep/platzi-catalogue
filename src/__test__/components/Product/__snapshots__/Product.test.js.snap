// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Product /> Should render snapshot 1`] = `
<WrapperComponent
  config={
    Object {
      "redux": undefined,
      "theme": true,
    }
  }
>
  <Component
    theme={
      Object {
        "borderRadius": Object {
          "button": "100px",
          "large": "18px",
          "medium": "8px",
          "none": "unset",
          "small": "3px",
          "total": "50%",
        },
        "boxShadows": Object {
          "main": "1px 2px 4px 0 rgba(21,99,157,.16)",
          "secondary": "0 0 0 1px #0791e6",
        },
        "breakpoints": Object {
          "desktop": "@media only screen and (min-width: 1280px)",
          "laptop": "@media only screen and (min-width: 1024px)",
          "mobile": "@media only screen and (min-width: 320px)",
          "screen": "@media only screen and (min-width: 1920px)",
          "smallMobile": "@media only screen and (min-width: 0px)",
          "tablet": "@media only screen and (min-width: 640px)",
        },
        "colors": Object {
          "bg": Object {
            "primary": "linear-gradient(90deg,#1c3643,#273b47 25%,#1e5372)",
            "quaternary": "#F3F8FB",
            "secondary": "white",
            "tertiary": "#0791e6",
          },
          "border": Object {
            "primary": "#A8B1C0",
            "secondary": "white",
          },
          "font": Object {
            "primary": "#131D30",
            "quaternary": "#A8B1C0",
            "secondary": "#98ca3f",
            "tertiary": "white",
          },
          "separator": Object {
            "primary": "#A8B1C0",
            "secondary": "#0791e6",
          },
        },
        "containerWidth": Object {
          "desktop": Object {
            "maxWidth": "85%",
            "width": "85%",
          },
          "fluid": Object {
            "maxWidth": "100%",
            "width": "100%",
          },
          "limitFluid": Object {
            "maxWidth": "1920px",
            "width": "100%",
          },
          "mobile": Object {
            "maxWidth": "92%",
            "width": "92%",
          },
          "screen": Object {
            "maxWidth": "75%",
            "width": "75%",
          },
          "smallMobile": Object {
            "maxWidth": "92%",
            "width": "92%",
          },
          "tablet": Object {
            "maxWidth": "85%",
            "width": "85%",
          },
        },
        "font": Object {
          "family": Object {
            "primary": "'Lato', sans-serif",
          },
          "letterSpacing": Object {
            "medium": "0.5rem",
            "normal": "0rem",
          },
          "lineHeight": Object {
            "medium": "0.6rem",
            "normal": "normal",
          },
          "size": Object {
            "h1": "3.75rem",
            "h2": "3rem",
            "h3": "2.5rem",
            "h4": "2.25rem",
            "h5": "1.625rem",
            "h6": "1.5rem",
            "help": "0.625em",
            "paragraph1": "1.125rem",
            "paragraph2": "1rem",
            "paragraph3": "0.875rem",
            "paragraph4": "0.75rem",
            "small": "0.5rem",
            "subtitle1": "1.1375rem",
            "subtitle2": "1.25rem",
            "subtitle3": "1.375rem",
          },
        },
        "name": "LIGHT",
        "transitions": Object {
          "button": "background-color 0.4s ease, color 0.4s ease, border-color 0.4s ease",
        },
        "utils": Object {
          "spacing": [Function],
        },
        "zIndex": Object {
          "backward": -1,
          "fixed": 2,
          "forward": 1,
          "higher": 3,
          "initial": 0,
          "super": 100,
        },
      }
    }
  >
    <Product
      hoverImage={
        Object {
          "alt": "Isotype",
          "src": "/icons/isotype.png",
        }
      }
      image={
        Object {
          "alt": "Isotype",
          "src": "/icons/isotype.png",
        }
      }
      name="name"
      onClick={[Function]}
      price={124}
      slug="slug"
    >
      <InternalLink
        href="/products/slug"
        onClick={[Function]}
        target="_blank"
      >
        <Component
          href="/products/slug"
          onClick={[Function]}
          passHref={false}
        >
          <div
            href="/products/slug"
            onClick={[Function]}
            passHref={false}
          >
            <a>
              <Productstyled__StyledProduct>
                <div
                  className="Productstyled__StyledProduct-sc-1btio9o-1 gWmVAk"
                >
                  <Productstyled__StyledProductImageContainer>
                    <div
                      className="Productstyled__StyledProductImageContainer-sc-1btio9o-0 ebPA-Dm"
                    >
                      <ProductImage
                        alt="Isotype"
                        className="bottom"
                        src="/icons/isotype.png"
                      >
                        <Image
                          alt="Isotype"
                          className="bottom"
                          layout="fill"
                          loading="lazy"
                          objectFit="cover"
                          quality={100}
                          src="/icons/isotype.png"
                        >
                          <div
                            style={
                              Object {
                                "bottom": 0,
                                "boxSizing": "border-box",
                                "display": "block",
                                "left": 0,
                                "margin": 0,
                                "overflow": "hidden",
                                "position": "absolute",
                                "right": 0,
                                "top": 0,
                              }
                            }
                          >
                            <img
                              alt="Isotype"
                              className="bottom"
                              decoding="async"
                              src="data:image/gif;base64,R0lGODlhAQABAIAAAAAAAP///yH5BAEAAAAALAAAAAABAAEAAAIBRAA7"
                              style={
                                Object {
                                  "border": "none",
                                  "bottom": 0,
                                  "boxSizing": "border-box",
                                  "display": "block",
                                  "height": 0,
                                  "left": 0,
                                  "margin": "auto",
                                  "maxHeight": "100%",
                                  "maxWidth": "100%",
                                  "minHeight": "100%",
                                  "minWidth": "100%",
                                  "objectFit": "cover",
                                  "objectPosition": undefined,
                                  "padding": 0,
                                  "position": "absolute",
                                  "right": 0,
                                  "top": 0,
                                  "visibility": "hidden",
                                  "width": 0,
                                }
                              }
                            />
                          </div>
                        </Image>
                      </ProductImage>
                      <ProductImage
                        alt="Isotype"
                        className="top"
                        src="/icons/isotype.png"
                      >
                        <Image
                          alt="Isotype"
                          className="top"
                          layout="fill"
                          loading="lazy"
                          objectFit="cover"
                          quality={100}
                          src="/icons/isotype.png"
                        >
                          <div
                            style={
                              Object {
                                "bottom": 0,
                                "boxSizing": "border-box",
                                "display": "block",
                                "left": 0,
                                "margin": 0,
                                "overflow": "hidden",
                                "position": "absolute",
                                "right": 0,
                                "top": 0,
                              }
                            }
                          >
                            <img
                              alt="Isotype"
                              className="top"
                              decoding="async"
                              src="data:image/gif;base64,R0lGODlhAQABAIAAAAAAAP///yH5BAEAAAAALAAAAAABAAEAAAIBRAA7"
                              style={
                                Object {
                                  "border": "none",
                                  "bottom": 0,
                                  "boxSizing": "border-box",
                                  "display": "block",
                                  "height": 0,
                                  "left": 0,
                                  "margin": "auto",
                                  "maxHeight": "100%",
                                  "maxWidth": "100%",
                                  "minHeight": "100%",
                                  "minWidth": "100%",
                                  "objectFit": "cover",
                                  "objectPosition": undefined,
                                  "padding": 0,
                                  "position": "absolute",
                                  "right": 0,
                                  "top": 0,
                                  "visibility": "hidden",
                                  "width": 0,
                                }
                              }
                            />
                          </div>
                        </Image>
                      </ProductImage>
                    </div>
                  </Productstyled__StyledProductImageContainer>
                  <Typography
                    as="h2"
                    className="typography"
                    color="secondary"
                    fontFamily="primary"
                    fontStyle="normal"
                    fontWeight="bold"
                    letterSpacing="normal"
                    lineHeight="normal"
                    textAlign="center"
                    textDecoration="none"
                    textTransform="none"
                    type="subtitle1"
                  >
                    <Typographystyled
                      as="h2"
                      className="typography"
                      color="secondary"
                      fontFamily="primary"
                      fontStyle="normal"
                      fontWeight="bold"
                      letterSpacing="normal"
                      lineHeight="normal"
                      textAlign="center"
                      textDecoration="none"
                      textTransform="none"
                      type="subtitle1"
                    >
                      <h2
                        className="Typographystyled-sc-1pbtdtz-0 TNqEh typography"
                        color="secondary"
                        fontFamily="primary"
                        fontStyle="normal"
                        fontWeight="bold"
                        letterSpacing="normal"
                        textDecoration="none"
                        type="subtitle1"
                      >
                        name
                      </h2>
                    </Typographystyled>
                  </Typography>
                  <Typography
                    as="h2"
                    className="typography"
                    color="primary"
                    fontFamily="primary"
                    fontStyle="normal"
                    fontWeight="bold"
                    letterSpacing="normal"
                    lineHeight="normal"
                    textAlign="center"
                    textDecoration="none"
                    textTransform="none"
                    type="subtitle1"
                  >
                    <Typographystyled
                      as="h2"
                      className="typography"
                      color="primary"
                      fontFamily="primary"
                      fontStyle="normal"
                      fontWeight="bold"
                      letterSpacing="normal"
                      lineHeight="normal"
                      textAlign="center"
                      textDecoration="none"
                      textTransform="none"
                      type="subtitle1"
                    >
                      <h2
                        className="Typographystyled-sc-1pbtdtz-0 hXKoMh typography"
                        color="primary"
                        fontFamily="primary"
                        fontStyle="normal"
                        fontWeight="bold"
                        letterSpacing="normal"
                        textDecoration="none"
                        type="subtitle1"
                      >
                        $ 124
                      </h2>
                    </Typographystyled>
                  </Typography>
                  <FancyLine
                    height="1px"
                    lineColor="secondary"
                  >
                    <FancyLinestyled
                      height="1px"
                      lineColor="secondary"
                    >
                      <div
                        className="FancyLinestyled-sc-1ffx9pt-0 gYjXDl"
                        height="1px"
                      />
                    </FancyLinestyled>
                  </FancyLine>
                  <FakeButton
                    centered={true}
                  >
                    <Fakebuttonstyled
                      centered={true}
                    >
                      <div
                        className="Fakebuttonstyled-gl3kvr-0 pZBOM"
                      >
                        <Typography
                          className="typography"
                          color="tertiary"
                          fontFamily="primary"
                          fontStyle="normal"
                          fontWeight="bold"
                          letterSpacing="normal"
                          lineHeight="normal"
                          textDecoration="none"
                          textTransform="none"
                          type="paragraph1"
                        >
                          <Typographystyled
                            as="p"
                            className="typography"
                            color="tertiary"
                            fontFamily="primary"
                            fontStyle="normal"
                            fontWeight="bold"
                            letterSpacing="normal"
                            lineHeight="normal"
                            textDecoration="none"
                            textTransform="none"
                            type="paragraph1"
                          >
                            <p
                              className="Typographystyled-sc-1pbtdtz-0 kQqjtw typography"
                              color="tertiary"
                              fontFamily="primary"
                              fontStyle="normal"
                              fontWeight="bold"
                              letterSpacing="normal"
                              textDecoration="none"
                              type="paragraph1"
                            >
                              Ver más
                            </p>
                          </Typographystyled>
                        </Typography>
                      </div>
                    </Fakebuttonstyled>
                  </FakeButton>
                </div>
              </Productstyled__StyledProduct>
            </a>
          </div>
        </Component>
      </InternalLink>
    </Product>
  </Component>
</WrapperComponent>
`;
